atlasdb-config/src/test/java/com/palantir/atlasdb/factory/timelock/ImmutableTimestampBridgingTimeLockServiceTest.java

```diff
@@ -0,0 +1,200 @@
+
+package com.palantir.atlasdb.factory.timelock;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Matchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyNoMoreInteractions;
+import static org.mockito.Mockito.when;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.UUID;
+import java.util.stream.IntStream;
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+import com.google.common.collect.Lists;
+import com.google.common.util.concurrent.RateLimiter;
+import com.palantir.atlasdb.http.errors.AtlasDbRemoteException;
+import com.palantir.lock.v2.IdentifiedTimeLockRequest;
+import com.palantir.lock.v2.LockImmutableTimestampResponse;
+import com.palantir.lock.v2.LockToken;
+import com.palantir.lock.v2.StartAtlasDbTransactionResponse;
+import com.palantir.lock.v2.TimelockService;
+
+public class ImmutableTimestampBridgingTimeLockServiceTest {
+    private static final AtlasDbRemoteException REMOTE_EXCEPTION_404 = mock(AtlasDbRemoteException.class);
+    private static final AtlasDbRemoteException REMOTE_EXCEPTION_503 = mock(AtlasDbRemoteException.class);
+    private static final Exception RUNTIME_EXCEPTION = new RuntimeException("boom");
+
+    private static final long FRESH_TIMESTAMP = 1L;
+    private static final LockImmutableTimestampResponse IMMUTABLE_TIMESTAMP_RESPONSE
+            = LockImmutableTimestampResponse.of(0L, LockToken.of(UUID.randomUUID()));
+    private static final StartAtlasDbTransactionResponse TRANSACTION_RESPONSE
+            = StartAtlasDbTransactionResponse.of(IMMUTABLE_TIMESTAMP_RESPONSE, FRESH_TIMESTAMP);
+
+    private static final IdentifiedTimeLockRequest IDENTIFIED_TIME_LOCK_REQUEST = IdentifiedTimeLockRequest.create();
+
+    private List<Exception> caughtExceptions = Lists.newArrayList();
+
+    private TimelockService timelockService = mock(TimelockService.class);
+    private TimelockService quickRetryingService
+            = new ImmutableTimestampBridgingTimeLockService(timelockService, RateLimiter.create(Double.MAX_VALUE));
+    private TimelockService standardBridgingService
+            = ImmutableTimestampBridgingTimeLockService.create(timelockService);
+
+    @Before
+    public void setUp() {
+        when(REMOTE_EXCEPTION_404.getStatus()).thenReturn(404);
+        when(REMOTE_EXCEPTION_503.getStatus()).thenReturn(503);
+
+        when(timelockService.getFreshTimestamp()).thenReturn(1L);
+        when(timelockService.lockImmutableTimestamp(any())).thenReturn(IMMUTABLE_TIMESTAMP_RESPONSE);
+    }
+
+    @After
+    public void verifyMocks() {
+        verifyNoMoreInteractions(timelockService);
+    }
+
+    @Test
+    public void defersToIndividualStepsIfStartTransactionNotAvailable() {
+        givenServerThrows404sOnStartTransaction();
+
+        whenStandardClientStartsTransactionNTimes(1);
```

Comment:
nit: method name is awkward, maybe better just `whenStandardClientStartsTransactionTimes`? Also goes for all the subsequent methods

Reply:
Alright, will remove the `N` before `Times` in most of these